rfcs/0050-verification-key-permissions.md

```diff
@@ -0,0 +1,81 @@
+
+## Motivation
+
+At the launch of the Berkeley network, the Mina Protocol does not yet guarantee backwards compatability of zkApps in future upgrades. Due to this, it is possible for an immutable contract to break -- any interactions with it that require proofs to be verified against it's verification key will no longer be accepted by the chain. Similarly, the on-chain features that the contract relied on may have been removed or modified.
+
+We wish for zkApps developers to be able to make their contract immutable without putting their contract at risk to breaking upon a future release, rendering any funds locked up in the contract inaccessible.
+
+## Detailed design
+
+NB: This RFC relies on the [protocol versioning RFC](TODO).
+
+In order to prevent a zkApp from breaking upon a future incompatible upgrade of the protocol, we will not allow users to set the `Impossible` or `Proof` permissions on the verification key. Instead, we will add 2 new permissions, `Impossible_during_hard_fork` and `Proof_during_hard_fork`, each of which will carry a specific transaction logic version to represent the hard fork they in which they are valid. Thus, the total set of permissions allowed to be set on verification keys would be:
+
+* `None`
+* `Either`
+* `Signature`
+* `Proof_during_hard_fork of txn_version`
+* `Impossible_during_hard_fork of txn_version`
+
+There is a restriction enforced in the transaction snark when a user attempts to set the permission to either `Proof_during_hard_fork` or `Impossible_during_hard_fork`. For any hard fork version, the transaction snark will only accept attempts to set one of these permissions to the current hard fork version. Any attempt to set one of these permissions to a hard fork version other than the current one will result in a failure. This failure should be implemented as a well formed transaction error rather than failing on chain, since it is something that can be checked statically before accepting transactions into the transaction pool. Thus, the transaction snark will fail to prove any such transactions, rather than failing on-chain.
+
+When the `txn_version` referenced by either of these new permissions matches the current hard fork version of the protocol, the permissions act exactly like their normal counterparts (`Proof` fields can only be updated with a valid proof, `Impossible` fields can never be updated). When the `txn_version` is older than (less than) the current hard fork version, then both of these permissions fallback to the `Signature` case, so that the now broken zkApps can be updated for the new hard fork.
+
+An account permission that is set to either of these new permissions can be updated when the `txn_version` referenced is older than the current hard fork version. This scenario ignores restrictions by `set_permission`, and operates as if `set_permission` were `Signature`.
```

Comment:
In the scenarios where the `txn_version` is older (less than) the current hard fork version, is there a risk in the zkApp's centralisation if it can only fall back to the `Signature` case? I suppose in some scenarios zkApp developers may allow users to participate in a vote, which could act as a `Signature` but might indeed be a `Proof`. Grateful for your thoughts!

Reply:
Unfortunately, if the release is not backwards compatible with the older proof system, we are not able to accept `Proof` authorizations since we cannot recursively verify it inside of the transaction snark. In the world where we don't have a backwards compatible upgrade, `Signature` is the only authorization which would still be valid.